### ISSUE TITLE
Inconsistent results from platform.machine() on Windows ARM64

### ISSUE THREAD

**Issue by cwalther** [python/cpython#98962]
# Bug report

Running x86 and AMD64 builds of Python 3.11.0 (official binary releases) in emulation on Windows 11 on an ARM64 machine gives me

```
Python 3.11.0 (main, Oct 24 2022, 18:26:48) [MSC v.1933 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import platform
>>> platform.machine()
'AMD64'
```

```
Python 3.11.0 (main, Oct 24 2022, 18:13:38) [MSC v.1933 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import platform
>>> platform.machine()
'ARM64'
```

This seem inconsistent. I would expect both to return `'ARM64'`, the actual underlying processor architecture, in the same way as running an x86 build on AMD64 returns `'AMD64'`. But even if they were reporting the emulated architecture, it should be `'AMD64'` and `'x86'`, not `'AMD64'` and `'ARM64'` (or possibly `'AMD64'` and `'AMD64'`, if x86 binaries are also handled by the AMD64 emulator, I have no idea how that works).

This makes it hard to distinguish between an AMD64 build running on ARM64 and an AMD64 build running on AMD64. In fact, the only `platform` function that gives any indication at all is `processor()`, and parsing the desired information out of the free-form text returned by that seems fiddly. (`'ARMv8 (64-bit) Family 8 Model 0 Revision   0,  '` and `'Intel64 Family 6 Model 44 Stepping 2, GenuineIntel'` here.)

Relatedly, how do I distinguish between an ARM64 build running on ARM64 and an AMD64 build running on ARM64? The inconsistency above makes them differ in `platform.machine()`, but what if that were fixed? This seems like a job for `platform.architecture()`, which is explicitly meant to give information about the running binary, but that just returns `('64bit', 'WindowsPE')` in both cases. `platform.python_compiler()` happens to have that information (`'MSC v.1933 64 bit (ARM64)'` vs. `'MSC v.1933 64 bit (AMD64)'`), but again that is fiddly free-form text and may not work for other compilers than MSC.

# Your environment

- CPython versions tested on: 3.11.0
- Operating system and architecture: Windows 11 10.0.22000 ARM64, Windows 10 10.0.19043 AMD64

**Comment by eryksun** [comment]
Out of curiosity, can you install the x86 and AMD64 versions of 3.12.0a1 and check `platform.machine()` in each? The new implementation in 3.12 prefers to get the CPU "Architecture" from WMI "Win32\_Processor" instead of from the system environment variables "PROCESSOR\_ARCHITEW6432" and "PROCESSOR\_ARCHITECTURE".

**Comment by cwalther** [comment]
Success! Thanks for the tip. In 3.12.0a1, all binaries return `platform.machine() == 'ARM64'` when run on ARM64 and `platform.machine() == 'AMD64'` when run on AMD64, which is what I would expect.

I notice however that, in a similar issue on macOS, #96993, @rossng seems to have a different expectation.

The problem remains that ARM64-on-ARM64 and AMD64-on-ARM64 are only distinguishable by `platform.python_compiler()`. Should I open a separate issue for that?

**Detailed output**
<details><summary>AMD64-on-ARM64</summary>

```
Python 3.12.0a1 (main, Oct 25 2022, 00:24:35) [MSC v.1933 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import platform
>>> for n in dir(platform):
...     try:
...             print("{:22s} {:s}".format(n, repr(getattr(platform, n)())))
...     except:
...             pass
...
_Processor             <platform._Processor object at 0x0000021960314F50>
_get_machine_win32     'ARM64'
_mac_ver_xml           None
_node                  'M1Win11'
_sys_version           ('CPython', '3.12.0a1', '', '', 'main', 'Oct 25 2022 00:24:35', 'MSC v.1933 64 bit (AMD64)')
_syscmd_ver            ('Microsoft', 'Windows', '10.0.22000')
architecture           ('64bit', 'WindowsPE')
java_ver               ('', '', ('', '', ''), ('', '', ''))
libc_ver               ('', '')
mac_ver                ('', ('', '', ''), '')
machine                'ARM64'
node                   'M1Win11'
platform               'Windows-11-10.0.22000-SP0'
processor              'ARMv8 (64-bit) Family 8 Model 0 Revision   0,  '
python_branch          ''
python_build           ('main', 'Oct 25 2022 00:24:35')
python_compiler        'MSC v.1933 64 bit (AMD64)'
python_implementation  'CPython'
python_revision        ''
python_version         '3.12.0a1'
python_version_tuple   ('3', '12', '0a1')
release                '11'
system                 'Windows'
uname                  uname_result(system='Windows', node='M1Win11', release='11', version='10.0.22000', machine='ARM64')
version                '10.0.22000'
win32_edition          'Professional'
win32_is_iot           False
win32_ver              ('11', '10.0.22000', 'SP0', 'Multiprocessor Free')
>>> ^Z
```
</details>
<details><summary>x86-on-ARM64</summary>

```
Python 3.12.0a1 (main, Oct 25 2022, 00:13:39) [MSC v.1933 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import platform
>>> for n in dir(platform):
...     try:
...             print("{:22s} {:s}".format(n, repr(getattr(platform, n)())))
...     except:
...             pass
...
_Processor             <platform._Processor object at 0x08BD6C78>
_get_machine_win32     'ARM64'
_mac_ver_xml           None
_node                  'M1Win11'
_sys_version           ('CPython', '3.12.0a1', '', '', 'main', 'Oct 25 2022 00:13:39', 'MSC v.1933 32 bit (Intel)')
_syscmd_ver            ('Microsoft', 'Windows', '10.0.22000')
architecture           ('32bit', 'WindowsPE')
java_ver               ('', '', ('', '', ''), ('', '', ''))
libc_ver               ('', '')
mac_ver                ('', ('', '', ''), '')
machine                'ARM64'
node                   'M1Win11'
platform               'Windows-11-10.0.22000-SP0'
processor              'ARMv8 (64-bit) Family 8 Model 0 Revision   0,  '
python_branch          ''
python_build           ('main', 'Oct 25 2022 00:13:39')
python_compiler        'MSC v.1933 32 bit (Intel)'
python_implementation  'CPython'
python_revision        ''
python_version         '3.12.0a1'
python_version_tuple   ('3', '12', '0a1')
release                '11'
system                 'Windows'
uname                  uname_result(system='Windows', node='M1Win11', release='11', version='10.0.22000', machine='ARM64')
version                '10.0.22000'
win32_edition          'Enterprise'
win32_is_iot           False
win32_ver              ('11', '10.0.22000', 'SP0', 'Multiprocessor Free')
>>> ^Z
```
</details>
<details><summary>ARM64-on-ARM64</summary>

```
Python 3.12.0a1 (main, Oct 25 2022, 00:13:29) [MSC v.1933 64 bit (ARM64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import platform
>>> for n in dir(platform):
...     try:
...             print("{:22s} {:s}".format(n, repr(getattr(platform, n)())))
...     except:
...             pass
...
_Processor             <platform._Processor object at 0x000001AAC8D78FB0>
_get_machine_win32     'ARM64'
_mac_ver_xml           None
_node                  'M1Win11'
_sys_version           ('CPython', '3.12.0a1', '', '', 'main', 'Oct 25 2022 00:13:29', 'MSC v.1933 64 bit (ARM64)')
_syscmd_ver            ('Microsoft', 'Windows', '10.0.22000')
architecture           ('64bit', 'WindowsPE')
java_ver               ('', '', ('', '', ''), ('', '', ''))
libc_ver               ('', '')
mac_ver                ('', ('', '', ''), '')
machine                'ARM64'
node                   'M1Win11'
platform               'Windows-11-10.0.22000-SP0'
processor              'ARMv8 (64-bit) Family 8 Model 0 Revision   0,  '
python_branch          ''
python_build           ('main', 'Oct 25 2022 00:13:29')
python_compiler        'MSC v.1933 64 bit (ARM64)'
python_implementation  'CPython'
python_revision        ''
python_version         '3.12.0a1'
python_version_tuple   ('3', '12', '0a1')
release                '11'
system                 'Windows'
uname                  uname_result(system='Windows', node='M1Win11', release='11', version='10.0.22000', machine='ARM64')
version                '10.0.22000'
win32_edition          'Professional'
win32_is_iot           False
win32_ver              ('11', '10.0.22000', 'SP0', 'Multiprocessor Free')
>>> ^Z
```
</details>
<details><summary>AMD64-on-AMD64</summary>

```
Python 3.12.0a1 (main, Oct 25 2022, 00:24:35) [MSC v.1933 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import platform
>>> for n in dir(platform):
...     try:
...             print("{:22s} {:s}".format(n, repr(getattr(platform, n)())))
...     except:
...             pass
...
_Processor             <platform._Processor object at 0x000002AF1B5FA630>
_get_machine_win32     'AMD64'
_mac_ver_xml           None
_node                  'DESKTOP-ILRVE2N'
_sys_version           ('CPython', '3.12.0a1', '', '', 'main', 'Oct 25 2022 00:24:35', 'MSC v.1933 64 bit (AMD64)')
_syscmd_ver            ('Microsoft', 'Windows', '10.0.19043')
architecture           ('64bit', 'WindowsPE')
java_ver               ('', '', ('', '', ''), ('', '', ''))
libc_ver               ('', '')
mac_ver                ('', ('', '', ''), '')
machine                'AMD64'
node                   'DESKTOP-ILRVE2N'
platform               'Windows-10-10.0.19043-SP0'
processor              'Intel64 Family 6 Model 44 Stepping 2, GenuineIntel'
python_branch          ''
python_build           ('main', 'Oct 25 2022 00:24:35')
python_compiler        'MSC v.1933 64 bit (AMD64)'
python_implementation  'CPython'
python_revision        ''
python_version         '3.12.0a1'
python_version_tuple   ('3', '12', '0a1')
release                '10'
system                 'Windows'
uname                  uname_result(system='Windows', node='DESKTOP-ILRVE2N', release='10', version='10.0.19043', machine='AMD64')
version                '10.0.19043'
win32_edition          'Professional'
win32_is_iot           False
win32_ver              ('10', '10.0.19043', 'SP0', 'Multiprocessor Free')
>>> ^Z
```
</details>
<details><summary>x86-on-AMD64</summary>

```
Python 3.12.0a1 (main, Oct 25 2022, 00:13:39) [MSC v.1933 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import platform
>>> for n in dir(platform):
...     try:
...             print("{:22s} {:s}".format(n, repr(getattr(platform, n)())))
...     except:
...             pass
...
_Processor             <platform._Processor object at 0x0163CE70>
_get_machine_win32     'AMD64'
_mac_ver_xml           None
_node                  'DESKTOP-ILRVE2N'
_sys_version           ('CPython', '3.12.0a1', '', '', 'main', 'Oct 25 2022 00:13:39', 'MSC v.1933 32 bit (Intel)')
_syscmd_ver            ('Microsoft', 'Windows', '10.0.19043')
architecture           ('32bit', 'WindowsPE')
java_ver               ('', '', ('', '', ''), ('', '', ''))
libc_ver               ('', '')
mac_ver                ('', ('', '', ''), '')
machine                'AMD64'
node                   'DESKTOP-ILRVE2N'
platform               'Windows-10-10.0.19043-SP0'
processor              'Intel64 Family 6 Model 44 Stepping 2, GenuineIntel'
python_branch          ''
python_build           ('main', 'Oct 25 2022 00:13:39')
python_compiler        'MSC v.1933 32 bit (Intel)'
python_implementation  'CPython'
python_revision        ''
python_version         '3.12.0a1'
python_version_tuple   ('3', '12', '0a1')
release                '10'
system                 'Windows'
uname                  uname_result(system='Windows', node='DESKTOP-ILRVE2N', release='10', version='10.0.19043', machine='AMD64')
version                '10.0.19043'
win32_edition          'Enterprise'
win32_is_iot           False
win32_ver              ('10', '10.0.19043', 'SP0', 'Multiprocessor Free')
>>> ^Z
```
</details>

**Comment by eryksun** [comment]
> The problem remains that ARM64-on-ARM64 and AMD64-on-ARM64

Have you checked `os.environ['PROCESSOR_ARCHITECTURE']`? That should be the architecture of the current process. 

Since you don't control the environment that was used to create the current process, it would be more reliable to call WinAPI `GetSystemInfo()` via PyWin32 or ctypes. This should return `wProcessorArchitecture` as `PROCESSOR_ARCHITECTURE_INTEL` (0) for x86, `PROCESSOR_ARCHITECTURE_AMD64` (9) for AMD64 (x64), and `PROCESSOR_ARCHITECTURE_ARM64` (12) for ARM64.

**Comment by cwalther** [comment]
Confirmed, these methods both work on all combinations, on both 3.11.0 and 3.12.0a1. Thanks! That will do as a workaround, although of course it would be nicer to have it abstracted in the `platform` module, ideally in a way that works identically on other OSes (although at the moment Windows is all I care about).

(It seems weird that two processes started from the same shell would see different environments, but then I’m used to weirdness from Microsoft.)

**Detailed output**

<details><summary>Script</summary>

```python
import os
print('PROCESSOR_ARCHITECTURE:', os.environ['PROCESSOR_ARCHITECTURE'])
import ctypes
import ctypes.wintypes
class SYSTEM_INFO(ctypes.Structure):
	_fields_ = [('wProcessorArchitecture', ctypes.wintypes.WORD), ('dwPageSize', ctypes.wintypes.DWORD), ('lpMinimumApplicationAddress', ctypes.wintypes.LPVOID), ('lpMaximumApplicationAddress', ctypes.wintypes.LPVOID), ('dwActiveProcessorMask', ctypes.c_void_p), ('dwNumberOfProcessors', ctypes.wintypes.DWORD), ('dwProcessorType', ctypes.wintypes.DWORD), ('dwAllocationGranularity', ctypes.wintypes.DWORD), ('wProcessorLevel', ctypes.wintypes.WORD), ('wProcessorRevision', ctypes.wintypes.WORD)]
info = SYSTEM_INFO()
ctypes.windll.kernel32.GetSystemInfo(ctypes.byref(info))
print('GetSystemInfo:      ', {f[0]: getattr(info, f[0]) for f in SYSTEM_INFO._fields_})
ctypes.windll.kernel32.GetNativeSystemInfo(ctypes.byref(info))
print('GetNativeSystemInfo:', {f[0]: getattr(info, f[0]) for f in SYSTEM_INFO._fields_})
```
</details>
<details><summary>AMD64-on-ARM64</summary>

```
PROCESSOR_ARCHITECTURE: AMD64
GetSystemInfo:       {'wProcessorArchitecture': 9, 'dwPageSize': 4096, 'lpMinimumApplicationAddress': 65536, 'lpMaximumApplicationAddress': 140737488289791, 'dwActiveProcessorMask': 15, 'dwNumberOfProcessors': 4, 'dwProcessorType': 8664, 'dwAllocationGranularity': 65536, 'wProcessorLevel': 15, 'wProcessorRevision': 1034}
GetNativeSystemInfo: {'wProcessorArchitecture': 9, 'dwPageSize': 4096, 'lpMinimumApplicationAddress': 65536, 'lpMaximumApplicationAddress': 140737488289791, 'dwActiveProcessorMask': 15, 'dwNumberOfProcessors': 4, 'dwProcessorType': 8664, 'dwAllocationGranularity': 65536, 'wProcessorLevel': 15, 'wProcessorRevision': 1034}
```
</details>
<details><summary>x86-on-ARM64</summary>

```
PROCESSOR_ARCHITECTURE: x86
GetSystemInfo:       {'wProcessorArchitecture': 0, 'dwPageSize': 4096, 'lpMinimumApplicationAddress': 65536, 'lpMaximumApplicationAddress': 2147418111, 'dwActiveProcessorMask': 15, 'dwNumberOfProcessors': 4, 'dwProcessorType': 586, 'dwAllocationGranularity': 65536, 'wProcessorLevel': 15, 'wProcessorRevision': 1034}
GetNativeSystemInfo: {'wProcessorArchitecture': 9, 'dwPageSize': 4096, 'lpMinimumApplicationAddress': 65536, 'lpMaximumApplicationAddress': 2147418111, 'dwActiveProcessorMask': 15, 'dwNumberOfProcessors': 4, 'dwProcessorType': 586, 'dwAllocationGranularity': 65536, 'wProcessorLevel': 15, 'wProcessorRevision': 1034}
```
</details>
<details><summary>ARM64-on-ARM64</summary>

```
PROCESSOR_ARCHITECTURE: ARM64
GetSystemInfo:       {'wProcessorArchitecture': 12, 'dwPageSize': 4096, 'lpMinimumApplicationAddress': 65536, 'lpMaximumApplicationAddress': 140737488289791, 'dwActiveProcessorMask': 15, 'dwNumberOfProcessors': 4, 'dwProcessorType': 0, 'dwAllocationGranularity': 65536, 'wProcessorLevel': 0, 'wProcessorRevision': 0}
GetNativeSystemInfo: {'wProcessorArchitecture': 12, 'dwPageSize': 4096, 'lpMinimumApplicationAddress': 65536, 'lpMaximumApplicationAddress': 140737488289791, 'dwActiveProcessorMask': 15, 'dwNumberOfProcessors': 4, 'dwProcessorType': 0, 'dwAllocationGranularity': 65536, 'wProcessorLevel': 0, 'wProcessorRevision': 0}
```
</details>
<details><summary>AMD64-on-AMD64</summary>

```
PROCESSOR_ARCHITECTURE: AMD64
GetSystemInfo:       {'wProcessorArchitecture': 9, 'dwPageSize': 4096, 'lpMinimumApplicationAddress': 65536, 'lpMaximumApplicationAddress': 140737488289791, 'dwActiveProcessorMask': 3, 'dwNumberOfProcessors': 2, 'dwProcessorType': 8664, 'dwAllocationGranularity': 65536, 'wProcessorLevel': 6, 'wProcessorRevision': 11266}
GetNativeSystemInfo: {'wProcessorArchitecture': 9, 'dwPageSize': 4096, 'lpMinimumApplicationAddress': 65536, 'lpMaximumApplicationAddress': 140737488289791, 'dwActiveProcessorMask': 3, 'dwNumberOfProcessors': 2, 'dwProcessorType': 8664, 'dwAllocationGranularity': 65536, 'wProcessorLevel': 6, 'wProcessorRevision': 11266}
```
</details>
<details><summary>x86-on-AMD64</summary>

```
PROCESSOR_ARCHITECTURE: x86
GetSystemInfo:       {'wProcessorArchitecture': 0, 'dwPageSize': 4096, 'lpMinimumApplicationAddress': 65536, 'lpMaximumApplicationAddress': 2147418111, 'dwActiveProcessorMask': 3, 'dwNumberOfProcessors': 2, 'dwProcessorType': 586, 'dwAllocationGranularity': 65536, 'wProcessorLevel': 6, 'wProcessorRevision': 11266}
GetNativeSystemInfo: {'wProcessorArchitecture': 9, 'dwPageSize': 4096, 'lpMinimumApplicationAddress': 65536, 'lpMaximumApplicationAddress': 4294901759, 'dwActiveProcessorMask': 3, 'dwNumberOfProcessors': 2, 'dwProcessorType': 8664, 'dwAllocationGranularity': 65536, 'wProcessorLevel': 6, 'wProcessorRevision': 11266}
```
</details>

**Comment by eryksun** [comment]
I just checked the [source](https://github.com/qemu/qemu/blob/ade760a2f63804b7ab1839fbc3e5ddbf30538718/linux-user/syscall.c#L10648-L10669) for QEMU user space emulation, which is used on Linux and BSD to emulate another CPU architecture in user space, relying on native system calls. When translating the `uname()` system call for an emulated process, it replaces the native machine with the emulated machine. Thus it seems that `platform.machine()` for an x64 Python that's running under QEMU emulation on ARM64 would report "x86_64" instead of "aarch64" (ARM64). In contrast, on Windows we return the native OS architecture. Should this be made consistent across platforms? If so, which choice is correct? If the choice to return the emulated architecture is correct, we could switch to calling `GetNativeSystemInfo()` on Windows, which will return the emulated architecture, instead of using a WMI query that returns the true machine architecture.

---

Off-topic comments on using ctypes: 

* I would include the `wReserved` field in `SYSTEM_INFO` instead of implicitly relying on alignment padding to set the offset of `dwPageSize`.
* It's better to use `kernel32 = ctypes.WinDLL('kernel32', use_last_error=True)` instead of `ctypes.windll.kernel32`. The `ctypes.windll` loader caches `WinDLL` instances, which cache function prototypes. This can cause conflicts between packages. Also, a separate `WinDLL` instance can set `use_last_error=True` to enable the use of reliable `ctypes.get_last_error()` instead of less reliable `kernel32.GetLastError()`.
* ctypes opens scripts up to problems with data loss and crashes due to memory and stack corruption. To help avoid this, in general it's better to define function prototypes such as `kernel32.GetSystemInfo.argtypes = (ctypes.POINTER(SYSTEM_INFO),)`. ctypes will type check arguments if `argtypes` is set, which helps to ensure that low-level C code is never passed bad arguments.

**Comment by cwalther** [comment]
I’m not sure what is _correct_. Maybe before deciding about that one should discuss what is _useful_. I have use cases for wanting to know
* code of what architecture(s) the machine is capable of running, so that when I’m installing software, I can install the version compiled for that architecture. This is what to me `platform.machine()` is documented to do, and does on Windows in 3.12.0a1: return something that depends _only_ on the machine (including installed OS) and _not_ on the running interpreter. However the documentation is not very clear, and apparently opinions differ on that (see #96993). An ideal function for this purpose would return a list of architectures in order of preference, maybe even marked by whether supported natively or in emulation.
* the architecture of the running interpreter, so that in a setuptools _setup.py_ I know what architecture we are building for (which wouldn’t necessarily need to be the same, but as far as I know there is no support for cross-compilation, so it is in practice, and I have not found any way of querying the latter directly) and can supply the corresponding libraries. This seems what `platform.architecture()` was intended to do, but does not on Windows (because it bins all architectures into the two categories “32bit” and “64bit”) nor on macOS (because it looks at the binary, not the process, and may choose wrong when the binary is fat).

So far I do not have a use case for wanting to know the emulated architecture, but it may exist.

I would consider consistency across platforms very desirable (and, at the moment, lacking). What would also be useful in that regard is if one could agree on a known set of possible values for an “architecture”, in particular choose one of “x86_64” and “AMD64” and one of “aarch64” and “ARM64” and stick to it.

(Thanks for the ctypes tips! Omitting `wReserved` was an oversight and I didn’t notice that my code was just working by accident due to alignment. I wasn’t aware of the second point, and left out the third in a quick and minimal script.)

**Comment by StanFromIreland** [comment]
cc @diegorusso

**Comment by zooba** [comment]
Since we're apparently going to discuss this issue again:
 
> code of what architecture(s) the machine is capable of running

Agreed, this is what `platform` should be returning. We might have to more clearly define it as "what native processes may be launched", since there are complications when it comes to emulation and loading dynamic modules/DLLs.

> the architecture of the running interpreter ...

This should be a `sys` property. It can be inferred from `sys.winver` on Windows, but probably `sys.implementation` should have it, since this property is only really descriptive of the interpreter and not _actually_ useful, because...

> the architecture of the running interpreter, so that in a setuptools setup.py I know what architecture we are building for

This is 100% a `sysconfig` property, with allowance for cross-compiling (that doesn't have to be built into `sysconfig`, but anyone _retrieving_ the property should allow their caller to override it).

I'm not totally sure the docs are clear on these, but that's the intent.

**Comment by diegorusso** [comment]
Apologies for the delay. I built main (3.15) for arm64, x64 and x86 on my Windows on Arm (on Parallels). I ran the script

``` python
import platform
import sys
import sysconfig


print("PLATFORM")
print("Architecture: ", platform.architecture())
print("Machine: ", platform.machine())
print("Processor: ", platform.processor())
print("uname: ", platform.uname())
print("Python Compiler: ", platform.python_compiler())

print("\nSYS")
print("Winver", sys.winver)
print("Implementation: ", sys.implementation)

print("\nSYSCONFIG")
print("Platform: ", sysconfig.get_platform())
```

and these are the results for:

### ARM64
```
PS C:\Users\dierus01\cpython> .\PCbuild\arm64\python.exe .\win_platform.py
PLATFORM
Architecture:  ('64bit', 'WindowsPE')
Machine:  ARM64
Processor:  ARMv8 (64-bit) Family 8 Model 0 Revision   0,
uname:  uname_result(system='Windows', node='DESKTOP-4LN91OF', release='11', version='10.0.22631', machine='ARM64')
Python Compiler:  MSC v.1943 64 bit (ARM64)

SYS
Winver 3.15-arm64
Implementation:  namespace(name='cpython', cache_tag='cpython-315', version=sys.version_info(major=3, minor=15, micro=0, releaselevel='alpha', serial=0), hexversion=51314848, supports_isolated_interpreters=True)

SYSCONFIG
Platform:  win-arm64
```

### x64
```
PS C:\Users\dierus01\cpython> .\PCbuild\amd64\python.exe .\win_platform.py
PLATFORM
Architecture:  ('64bit', 'WindowsPE')
Machine:  ARM64
Processor:  ARMv8 (64-bit) Family 8 Model 0 Revision   0,
uname:  uname_result(system='Windows', node='DESKTOP-4LN91OF', release='11', version='10.0.22631', machine='ARM64')
Python Compiler:  MSC v.1943 64 bit (AMD64)

SYS
Winver 3.15
Implementation:  namespace(name='cpython', cache_tag='cpython-315', version=sys.version_info(major=3, minor=15, micro=0, releaselevel='alpha', serial=0), hexversion=51314848, supports_isolated_interpreters=True)

SYSCONFIG
Platform:  win-amd64
```

### Win32
```
PLATFORM
Architecture:  ('32bit', 'WindowsPE')
Machine:  ARM64
Processor:  ARMv8 (64-bit) Family 8 Model 0 Revision   0,
uname:  uname_result(system='Windows', node='DESKTOP-4LN91OF', release='11', version='10.0.22631', machine='ARM64')
Python Compiler:  MSC v.1943 32 bit (Intel)

SYS
Winver 3.15-32
Implementation:  namespace(name='cpython', cache_tag='cpython-315', version=sys.version_info(major=3, minor=15, micro=0, releaselevel='alpha', serial=0), hexversion=51314848, supports_isolated_interpreters=True)

SYSCONFIG
Platform:  win32
```

This seems consistent independently on how you compile CPython: the machine is always ARM64 as well as the processor.
The `sysconfig.get_platform()` though gives another view which doesn't align with what reported by uname/machine.

The `sysconfig` module is all about CPython (and not the system): the way I read the get_platform() is that the CPython binary is built for a specific platform and not the running platform.

**Comment by zooba** [comment]
> The `sysconfig` module is all about CPython (and not the system): the way I read the get_platform() is that the CPython binary is built for a specific platform and not the running platform.

Correct. `sysconfig` is what you need to know to build CPython/extensions, `platform` is what humans expect to know about the hardware, and `sys` is what you need to know to _run_ Python (and indirectly through `sys.meta_path`, `importlib` knows how to locate compatible binaries at runtime). There may be some amount of data shared between them, but the intents are clear enough, so when one is inconsistent with another, you need to look at what the module is trying to provide and determine whether it's consistent with _that_.

**Comment by cwalther** [comment]
For a complete picture we would now need to see the output of the same binaries when running on an AMD64 machine, but so far this all looks good to me.

As far as I can tell, my original issue is fixed. @zooba’s explanations of how things are supposed to work match my expectations. My mentioned two use cases are covered by `platform.machine()` and `sysconfig.get_platform()`. In that regard, this issue could be closed, as far as I am concerned.

However, I have in the meantime encountered a third use case (two years ago already, so I don’t remember the details of what research I did or did not do at the time): Needing to know what directory to pass to `os.add_dll_directory()` to load DLLs via ctypes, because our DLLs (and their dependencies) for ARM64, AMD64 and x86 are in different directories. For that, I need to know the architecture of the running interpreter. I am currently calling `kernel32.GetSystemInfo()` via ctypes for that, as suggested by @eryksun. That works, so I am not going to change it, but I would still be interested in what the recommended solution is, out of curiosity. It seems like `sys.winver` would be the one to use here, although it is documented to be for a completely different purpose, so I am not sure how guaranteed that is to stay reliable. Or would it be `sysconfig.get_platform()`, even though this is not about building extensions? As far as I can tell, both would work, but what is the recommendation here? Is there one, or is this still a missing piece, so this issue should be kept open?

(I am now wondering whether I could just always call `os.add_dll_directory()` on all three directories, and Windows would automatically take the correct one of the three found DLLs, much as we do with $PATH, but I have not tried that.)

**Comment by zooba** [comment]
> It seems like `sys.winver` would be the one to use here, although it is documented to be for a completely different purpose, so I am not sure how guaranteed that is to stay reliable.

Technically it can change between feature releases without warning, but that's pretty unlikely to happen. I agree it's not the ideal way to detect the running process.

Strictly speaking, `importlib.machinery.EXTENSION_SUFFIXES` is the right source, but that's harder to handle than `sys.winver` and less obvious. We're currently working up a new `sys.abi_info` that is probably where we should capture it.  I linked your comment from https://discuss.python.org/t/94422/ to see whether it makes sense.

But until then, `sys.winver` basically captures the distinct ABIs on Windows, so it'll be safe enough to rely on it here.

**Comment by encukou** [comment]
> ARM64, AMD64 and x86 

Isn't that `platform.machine()`?

**Comment by zooba** [comment]
No, because Windows can emulate other architectures. So the platform might be one architecture, but the Python runtime is a different one. Sometimes you need to choose whether you want to know what the OS is (`platform`) and sometimes you need compatibility with the current process (`sys`/`sysconfig`).

**Comment by zooba** [comment]
Having it match `platform.machine()`'s set of return values (as far as is possible, which isn't far, but that's the point) would be reasonable. So in that sense, it's the same function, but the source of the info is the build process (what `platform.machine()` was targeted when building this runtime) rather than the current machine.

**Comment by isuruf** [comment]
On other systems like Linux and macOS, `platform.machine()` gives the processor type of the emulated processor (and not the native processor) when using QEMU or Rossetta2 respectively. I don't see why windows has to diverge in that sense.

**Comment by zooba** [comment]
> I don't see why windows has to diverge in that sense.

Because it doesn't emulate a processor - it chooses a subsystem each time you launch an executable, and that subsystem will either reconfigure the processor or convert the binary.

So `platform.machine()` tells you what kind of executable runs natively (and implicitly, that tells you what kinds will run emulated), while `sysconfig.get_platform()` tells you what kind of DLL would load into the current running process (which is more broadly applicable information, but I'm offering easily testable definitions). The only two current options for the former are 64bit and ARM64, as those are the two architectures supported by Windows right now, and the third option for the latter is 32bit because that subsystem is still available (on both architectures).

A proposal for Windows to "not diverge" would be some kind of lie about what architectures are supported, and the key design criteria for the `platform` module is that we should tell the truth about the hardware rather than lying. Right now, we do what the `platform` module is designed for - if other OS's are lying about their hardware, then fixing the divergence would involve changing them, and since that wouldn't help you here, it means we need to find an alternative option for you (which I suspect is `sysconfig.get_platform()`, since that's practically always the right answer for people who discover this nuance of the `platform` module).

**Comment by cwalther** [comment]
Perhaps a clarification of the documentation would be in order to avoid such diverging interpretations in the future.

@isuruf, how does one query the actual hardware architecture of the machine on Linux and macOS, if that is not what `platform.machine()` does?

**Comment by zooba** [comment]
> Perhaps a clarification of the documentation would be in order to avoid such diverging interpretations in the future.

Agreed. The [platform.machine](https://docs.python.org/3/library/platform.html#platform.machine) docs could do with repeating the page title ("the underlying platform's identifying data") and probably should cross-reference [sysconfig.get_platform](https://docs.python.org/3/library/sysconfig.html#sysconfig.get_platform), which might also benefit from a clear "this is the platform the current Python runtime was built for, and any compatible modules should also be built for this platform".

**Comment by isuruf** [comment]
> @isuruf, how does one query the actual hardware architecture of the machine on Linux and macOS, if that is not what platform.machine() does?

You usually don't. If a user is using x86_64 python on a Apple Silicon machine, then everything you use from that python process is x86_64 too. You can query using `sysctl -n machdep.cpu.brand_string` to get `Apple M4`, but that's about it. Everything else reports x86_64.

> Because it doesn't emulate a processor - it chooses a subsystem each time you launch an executable, and that subsystem will either reconfigure the processor or convert the binary.

This is true for macos too. From a x86_64 process you can launch a arm64 process by using something `arch -arm64 /path/to/fat/executable` if the executable has both arm64 and x86_64 parts in it and just `/path/to/arm64/executable` if it has only the arm64 slice. I guess this is the same for windows where x86_64 python can launch arm64 binaries, but just not load arm64 only libraries to the process.

**Comment by isuruf** [comment]
One use-case that I want to highlight is the ability to run an application that was compiled for x86_64 on a arm64 system without any changes being needed. When `platform.machine()` reports that it is `arm64`, then the `x86_64` application will take a different code path or error out depending on how `platform.machine()` is being used.

For eg:
- `conda install numpy` fails right now because `conda` is a x86_64 executable, but tries to install a arm64 `numpy` into the x86_64 environment)
- `onnxruntime` will try to load arm64 DLLs because of https://github.com/microsoft/onnxruntime/blob/11e2f5c14a1e063a2e707eec5762ff52dd7b1328/onnxruntime/__init__.py#L221-L227 (Guesswork. Haven't tried it)

**Comment by zooba** [comment]
> When `platform.machine()` reports that it is `arm64`, then the `x86_64` application will take a different code path or error out depending on how `platform.machine()` is being used.

Well, that'd be bugs. `onnxruntime` especially is clearly a bug (and proof that I don't actually review all the Python code that Microsoft puts out 😉 ). They should be using `sysconfig` because they want a library that's compatible with the current process.

The conda case is more challenging, because it seems reasonable to be able to use an x86_64 conda to create an arm64 environment, just as it could be totally reasonable for an arm64 conda to create an x86_64 environment (e.g. to use an un-ported dependency). This is a case where I'd encourage the user to specify, regardless of whether using `platform` or `sysconfig` for the default - either is justifiable here, IMHO. (I've been facing the same question with the Python install manager, which is _not_ ARM64-native, but is going to install/launch ARM64 by default on that machine when we flip the switch. Very overridable by users, but will default to the machine because I don't want people thinking about how the install tool was implemented.)

**Comment by cwalther** [comment]
> > how does one query the actual hardware architecture of the machine on Linux and macOS, if that is not what platform.machine() does?
> 
> You usually don't.

Well, _I_ do, as I’ve described further up. One question I need to answer is “I have executables for different architectures, which one of them should I try to run or install preferably?” Granted, that question does not usually occur on macOS because it has fat binaries, but Windows and Linux don’t.

**Comment by isuruf** [comment]
> They should be using sysconfig because they want a library that's compatible with the current process.

Yes, but 99.9% (no hard-data) of use-case for `platform.machine()` is for this which means we have to patch out `99.9%` of use-cases to support the behaviour of the rare use-case. Adding a separate API for that use-case is the best option IMHO. It'd be hard to get it right for Linux, but should be doable for macOS and windows. My understanding previously was that the times where `sysconfig.get_platform()` is useful was when we are building software (i.e. cross compiling), but for library things like `onnxruntime/__init__.py`, it was okay to use `platform.machine()`.

> Granted, that question does not usually occur on macOS because it has fat binaries,

Yes, but that's only for system binaries. Most software don't provide fat binaries.
